### PR TITLE
SLE-125, SLE-129 Analysis of changed files

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/command/AnalyzeChangeSetCommand.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    final Set<IProject> selectedProjects = new HashSet<>();
+
+    @SuppressWarnings("rawtypes")
+    List elems = selection.toList();
+    for (Object elem : elems) {
+      if (elem instanceof IResource) {
+        selectedProjects.add(((IResource) elem).getProject());
+      } else if (elem instanceof IAdaptable) {
+        IResource res = (IResource) ((IAdaptable) elem).getAdapter(IResource.class);
```
Comment (SonarTech) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/command/AnalyzeChangeSetCommand.java]: MINOR Remove this unnecessary cast to "IResource". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1905)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeChangedFilesJob.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        collect(subscriber, children[i], changedFiles);
+      }
+    } catch (TeamException e) {
+      throw new IllegalStateException("Unable to analyze changed files", e);
+    }
+    return changedFiles;
+  }
+
+  void collect(Subscriber subscriber, IResource resource, Collection<IFile> changedFiles) throws TeamException {
+    IFile file = (IFile) resource.getAdapter(IFile.class);
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeChangedFilesJob.java]: MINOR Remove this unnecessary cast to "IFile". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1905)

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/views/issues/ChangeSetIssuesView.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  private static void registerJobListener(Job job) {
+    job.addJobChangeListener(new JobChangeAdapter() {
+      @Override
+      public void done(IJobChangeEvent event) {
+        Display.getDefault().asyncExec(() -> {
+          // Enable button
+          ChangeSetIssuesView.instance.btn.setEnabled(true);
+          if (Status.OK_STATUS == event.getResult()) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/views/issues/ChangeSetIssuesView.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/views/issues/ChangeSetIssuesView.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  public static final String ID = SonarLintUiPlugin.PLUGIN_ID + ".views.issues.ChangeSetIssuesView";
+  private static LocalDateTime lastRefresh;
+  private static ChangeSetIssuesView instance;
+  private Label label;
+  private Composite bottom;
+  private Button btn;
+
+  public ChangeSetIssuesView() {
+    super(SonarLintUiPlugin.PLUGIN_ID + ".views.issues.changeSetIssueMarkerGenerator");
+    instance = this;
```
Comment (SonarTech) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/views/issues/ChangeSetIssuesView.java]: CRITICAL Synchronize this lazy initialization of 'instance' [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2444)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* CRITICAL 1 critical
* MAJOR 2 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeChangedFilesJob.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    super("Analyze changeset");
+    this.projects = projects;
+  }
+
+  @Override
+  protected IStatus run(IProgressMonitor monitor) {
+    projects.stream().forEach(MarkerUtils::deleteChangeSetIssuesMarkers);
+    List<Job> jobs = new ArrayList<>();
+    for (IProject project : projects) {
+      AnalyzeProjectRequest req = new AnalyzeProjectRequest(project, collectChangedFiles(project, monitor), TriggerType.CHANGESET);
```
Comment (dbmeneses) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeChangedFilesJob.java]: maybe if collectChangedFiles returns an empty list, simply `continue`?

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeChangedFilesJob.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+      AnalyzeProjectJob job = new AnalyzeProjectJob(req);
+      job.schedule();
+      jobs.add(job);
+    }
+    jobs.stream().forEach(j -> {
+      try {
+        j.join();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
```
Comment (dbmeneses) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeChangedFilesJob.java]: Do we need to do this?

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeChangedFilesJob.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      }
+    });
+    return Status.OK_STATUS;
```
Comment (dbmeneses) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeChangedFilesJob.java]: there should be a catch here, I think (and return Status warning or error). For example, collectFiles can return an IllegalStateException

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeChangedFilesJob.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        AnalyzeProjectRequest req = new AnalyzeProjectRequest(project, collectChangedFiles, TriggerType.CHANGESET);
+        AnalyzeProjectJob job = new AnalyzeProjectJob(req);
+        job.schedule();
+        jobs.add(job);
+      }
+    }
+    jobs.stream().forEach(j -> {
+      try {
+        j.join();
+      } catch (InterruptedException e) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeChangedFilesJob.java]: MAJOR Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2142)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* CRITICAL 1 critical
* MAJOR 3 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

**Pull request merged**